7 module complete

alexander-shvetsov opened the pull request.

alexander-shvetsov committed:
7 module complete

ios-review commented on 7 module/7.7 Tab Bar Controller/Model/Background.swift:
```diff
@@ -0,0 +1,30 @@
+}
+
+// метод сопоставления
+func getBackground(from buttonName: String) -> UIColor {
```
global func?

Будет ок только, если используется в сугубо технических целях, вроде log.

ios-review commented on 7 module/7.7 Tab Bar Controller/Model/Background.swift:
```diff
@@ -0,0 +1,30 @@
+
+import UIKit
+
+enum Background: String {
```
Сделать функцию вычисления UIColor в качестве функции-члена данного перечисления.

Затем создавать перечисление по rawValue и преобразовывать в UIColor.

```
enum Background: String {
    case green
    
    var color: UIColor {
          return .green
    }
}
```

alexander-shvetsov replied:
Реализовал общую модель для работы с цветом: как текстовой части, так и фона.

ios-review commented on 7 module/7.7 Tab Bar Controller/Controllers/ColorViewController.swift:
```diff
@@ -0,0 +1,33 @@
+
+import UIKit
+
+class ColorViewController: UIViewController {
```
Ошибка лейаута для кнопки изменить.
На SE (2nd)

<img width="389" alt="Screenshot 2020-09-24 at 12 39 27" src="https://user-images.githubusercontent.com/62285448/94105290-2984ae00-fe63-11ea-82c7-34194da66a0c.png">

alexander-shvetsov replied:
Так точно, не углядел. Кое-где произвёл рефакторинг UI

ios-review commented on 7 module/7.7 Tab Bar Controller/Controllers/BackgroundViewController.swift:
```diff
@@ -0,0 +1,40 @@
+    @IBAction func backgroundSetButton(_ sender: UIButton) {
+        
+        // получаем идентификатор кнопки
+        guard let isPressedButton = sender.currentTitle else { return }
```
currentTitle - не надежный идентификатор.
Title - это контент для пользователя.

Он может быть изменен в любое время другим разработчиком по просьбе контент-менеджера или аналитика.
Причин его изменения очень много.

В качестве идентификатора рекомендую использовать tag.
https://developer.apple.com/documentation/uikit/uiview/1622493-tag
tag можно задать в сториборде.

alexander-shvetsov replied:
Вот это да, совсем не подумал о пользовательской части. Действительно использовать tag гораздо выгоднее. Скорее всего для модели калькулятора этот подход и буду использовать.

ios-review commented on 7 module/7.7 Tab Bar Controller/Controllers/BackgroundViewController.swift:
```diff
@@ -0,0 +1,40 @@
+}
+
+/* расширяем класс до получения протокола делегата,
+который обязует контроллер, принимать новое значение фона */
```
Комментарии - топ😍

ios-review commented on 7 module/7.7 Tab Bar Controller/Controllers/ChangeColorViewController.swift:
```diff
@@ -0,0 +1,43 @@
+        colorTextField.text = settedColor
+    }
+
+    @IBAction func buttonPressed(_ sender: UIButton) {
```
Предлагаю поисследовать вопрос: "Как именовать @IBAction функции?"

После исследования написать краткое резюме + ссылки, которые были изучены.
Ответ написать к этому комментарию.

alexander-shvetsov replied:
Материалы:
https://swift.org/documentation/api-design-guidelines/#naming
https://medium.com/better-programming/naming-conventions-in-swift-4b7ca5eed4d2
Кое-что было написано в книгах...
===
That's my fail 😕
Всегда старался именовать всё отвечая на вопросы: что, для чего, что будет и т.д., а здесь почему-то отнёсся к кнопке чисто технически, по моему где-то увидел у кого-то такое именование, и "прилипло"... Моя ошибка.
Где-то читал, что код swift прекрасен как раз тем, что зная английский язык, в идеале, можно читать как рассказ. К этому стоит стремиться наверное ☺️

ios-review replied:
👍

ios-review commented on 7 module/7.7 Tab Bar Controller/Controllers/ChangeColorViewController.swift:
```diff
@@ -0,0 +1,43 @@
+        // передаём значение выбранного цвета
+        delegate?.setColor(settedColor)
+        
+        dismiss(animated: true, completion: nil) // close
```
`dismiss(animated: true)`

ios-review replied:
если что-то можно не делать, лучше этого не делать)

alexander-shvetsov replied:
А что, так можно было чтоли? (с)
😁
Так точно, буду знать. Благодарю!

ios-review commented on 7 module/7.7 Tab Bar Controller/Controllers/ColorViewController.swift:
```diff
@@ -0,0 +1,33 @@
+     наследуем новое значение тайтла, при выборе нового цвета */
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        guard let changeColorVC = segue.destination as? ChangeColorViewController else { return }
+        changeColorVC.settedColor = titleLabel.text!
```
force unwrap

alexander-shvetsov replied:
Виноват, исправился.

ios-review commented on 7 module/7.7 Tab Bar Controller/Model/Colors.swift:
```diff
@@ -0,0 +1,30 @@
+
+import UIKit
+
+enum Colors: String {
```
Две по сути очень близких модели в одном приложении, которое все про цвет.

Может воспользоваться одной моделью?
И добавить еще одно вычисляемое свойство для текстового описания цвета.

alexander-shvetsov replied:
Так точно, реализовал одну общую модель для работы с цветом.

ios-review reviewed (changes requested):
Упс.

Не нажал submit review)

alexander-shvetsov committed:
Work on bugs done

Реализовал общую модель по работе с цветом, как текста, так и фона.
Убрал небезопасное извлечение значений.
Мелкий рефакторинг кода.
Кое-где подправил UI.

ios-review commented on 7 module/7.7 Tab Bar Controller/Model/Colors.swift:
```diff
@@ -8,23 +8,32 @@
-
-// метод сопоставления
-func getColor(from buttonName: String) -> String {
+enum Colors: Int {
```
Мне кажется, с данным перечислением все стало гораздо круче)
Что вы думаете?

alexander-shvetsov replied:
Однозначно!

alexander-shvetsov replied:
Только почему-то в review отображён пример старого кода 🤔

ios-review commented on 7 module/7.7 Tab Bar Controller/Model/Colors.swift:
```diff
@@ -8,23 +8,32 @@
-    guard let colorType = Colors(rawValue: buttonName) else {
-        return "Выбран зелёный"
+    var background: UIColor {
+        switch self.rawValue {
```
если делать switch по self без rawValue, то тогда можно будет не указывать default блок.

ios-review commented on 7 module/7.7 Tab Bar Controller/Model/Colors.swift:
```diff
@@ -8,23 +8,32 @@
-    case .red: return "Выбран красный"
-    case .blue: return "Выбран синий"
+    var label: String {
+        switch self.rawValue {
```
если делать switch по self без rawValue, то тогда можно будет не указывать default блок.

ios-review reviewed (changes requested):
Стало очень круто!

Упростите реализацию свойств перечисления Colors?

alexander-shvetsov committed:
refactoring color model

ios-review reviewed (approved):
Все очень круто! Идеально!
Коротко, лакончино, с комментариями!
Молодец🔥

Мрежим.
